src/common/snippets/src/pass/insert_loops.cpp

```diff
@@ -0,0 +1,97 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <snippets/itt.hpp>
+#include "snippets/pass/insert_loops.hpp"
+#include "snippets/op/loop_helpers.hpp"
+
+#include <ngraph/rt_info.hpp>
+
+ngraph::snippets::pass::InsertLoops::InsertLoops(ov::PartialShape master_shape, size_t loop_depth, size_t vector_size)
+: m_master_shape(std::move(master_shape)), m_loop_depth(loop_depth), m_vector_size(vector_size) {
+    if (m_master_shape.size() < m_loop_depth)
+        throw ngraph_error("InsertLoops can't insert loops: master shape rank is too small");
+}
+
+bool ngraph::snippets::pass::InsertLoops::run_on_model(const std::shared_ptr<ov::Model> &model) {
+    RUN_ON_FUNCTION_SCOPE(InsertLoops);
+    if (m_master_shape.is_dynamic())
+        throw ngraph_error("InsertLoops doesn't support dynamic shapes yet");
+
+    const auto inner_dim = m_master_shape.size() - 1;
+    // Note: outer_dim will not be used if m_master_shape.size() < 2
+    const auto outer_dim = m_loop_depth == 2 ? m_master_shape.size() - 2 : -1;
+    const auto inner_work_amount = m_master_shape[inner_dim].get_length();
+    const auto outer_work_amount = m_loop_depth == 2 ? m_master_shape[outer_dim].get_length() : 1;
+
+    ParameterVector commonParams = model->get_parameters();
+    // Note that topological sort parses node arguments in reversed order, but results are added  - in direct order
+    // So ve need to pass the reversed results to LoopEnd to keep the original traversal order in topological sorter
+    const auto& orig_results = model->get_results();
+    ResultVector commonResults(orig_results.rbegin(), orig_results.rend());
+    std::vector<PartialShape> ioShapes;
+
+    const auto& body_rt_info = model->get_rt_info();
+    const auto& plugin_shapes = body_rt_info.find("PluginShapesOverride");
```

Comment:
I'm not sure that the current implementation is scalable for cases with several loops. Need to discuss

Reply:
Discussed, summary: it's ok since cases with several loops are not reshapable anyway.